### PR TITLE
feat(rabbitmq): error behaviour for replying error

### PR DESCRIPTION
**Pull request by perf2711**
I need my RPC service to return me errors, in case something goes wrong. Thus, a new error behavior has been added: 
```
  /**
   * If an exception occurs while handling the message, the error will be serialized and published on the `replyTo` queue.
   * If `replyTo` is not provided, the message will be NACKed without requeueing.
   * If publish fails, message will be NACKed and requeued.
   */
  REPLYERRORANDACK,
```

If this is set, the handler will behave as it is described in the comment.

Additionaly, I added a `try/catch` on JSON message parsing, and allowed `undefined` messages to go into the service. I believe that the service should be able to determine what to do with this. As of before, there was no option to publish an empty message to the service (`JSON.parse` would throw an error). This was the push for me to create the error behavior, as the service would by default endlessly requeue the message without me knowing what is going on.

### CODE REVIEW

**Review by Arebisc — COMMENTED** [review]
Looks fine. Thanks @perf2711

**Comment by WonderPanda** [comment]
Hey @perf2711 thanks for this contribution. One thing that I would definitely like to see before this could be merged is some additions to the integration test suites that validate the expected behavior. Let me know if you need some help getting pointed in the right direction

**Comment by perf2711** [comment]
@WonderPanda sure, I'll try to write some tests for this.
To be sure, you mean the e2e tests in `integration/rabbitmq/e2e`?

**Comment by WonderPanda** [comment]
@perf2711 Yes, you can use the included docker-compose file to get a running instance of RabbitMQ and then use `yarn test:integration`

**Comment by WonderPanda** [comment]
Hey @perf2711 I've been looking at this a little bit more and I can understand your desire for this solution but I think it might be possible to implement the ability for you to handle the errors in this way in a more generic fashion. Instead of adding a new error handling behavior at the connection level what if we added the option to supply your own error handling logic at the decorator level so that people can flexibly handle error cases differently across different queues/exchanges etc.
I'm wondering about adding in a new option method to the decorator that might look something like this:

```
errorHandler?: (
      connection: AmqpConnection,
      channel: amqplib.Channel,
      msg: amqplib.ConsumeMessage
    ) => void | Promise<void> 
```

That way if an error happens and an errorHandler was provided the current connection, channel and message that caused the error will be provided to your error handler which can do whatever it wants with this information. That would allow the community to supply additional plugins that cover difference scenarios and desired behavior without needing to bloat the core AmqpConnection code with more and more Error Handling Behaviors.

Tracking this as an enhancement issue here #127

**Comment by perf2711** [comment]
Hmm, that is a neat idea. To be honest, I don't know the NestJS internals that much, but could we extend it to guards, or injectable handlers? Looking to it now.

**Comment by perf2711** [comment]
By the way, I pushed some integration tests to the previous feature. I don't mind them being changed, if we go further with error handlers 😊

**Comment by WonderPanda** [comment]
Hey @perf2711 sorry for the delay on this, I've had a few other open source packages that I'm working on implementing and publishing from the monorepo so this has been on the backburner for me. I would really like to go with the more flexible approach so that the community can plugin their own customized error logic as opposed to baking in more named error handling strategies

**Comment by perf2711** [comment]
@WonderPanda I added the callback functionality. See if it complies with your idea 😊

File: packages/rabbitmq/src/amqp/connection.ts
Comment (WonderPanda) [inline packages/rabbitmq/src/amqp/connection.ts:421]: I feel like at this point we should think about removing this finally case and replacing the switch case with new named Error Handling Functions that represent the old behaviors, similar to the ReplyErrorCallback that you made. If we run the `msgOptions.errorCallbacks` as well as one from the legacy behaviors here we we will likely try to ack or nack a message that has already been acked or nacked which could lead to more problems

File: packages/rabbitmq/src/rabbitmq.interfaces.ts
Comment (WonderPanda) [inline packages/rabbitmq/src/rabbitmq.interfaces.ts:49]: I like the intent for `allowNonJsonMessages` but I think that it doesn't belong. One of the follow up things I'd like to do with the RabbitMQ library is provide pluggable serialization/deserialization of messages instead of assuming JSON on the users's behalf. For now I'd rather just let the JSON parsing throw an error and leave it to the errorCallback to decide what to do with the error

**Review by WonderPanda — CHANGES_REQUESTED** [review]

File: packages/rabbitmq/src/rabbitmq.interfaces.ts
Comment (WonderPanda) [inline packages/rabbitmq/src/rabbitmq.interfaces.ts:52]: The signature should be this:
```
IMessageErrorCallback = (
      connection: AmqpConnection,
      channel: amqplib.Channel,
      msg: amqplib.ConsumeMessage,
      error: any
    ) => void | Promise<void> 
```
so that the user has access to the AmqpConnection instance when they're implementing error handling. That way they can optionally publish messages or do other more advanced stuff.

I think we should restrict return to `void | Promise<void>` to prevent them from trying to return a value from the error handling function because nothing would be done with the return object

**Comment by WonderPanda** [comment]
@perf2711 Thanks for your persistence with this! I'm starting to get excited about the direction we're taking things the API is really shaping up. I left some feedback for you to check out, hopefully it's not annoying I know I'm picky sometimes

I really appreciate you taking the time to help put together a significant contribution for the library

File: packages/rabbitmq/src/amqp/connection.ts
Comment (perf2711) [reply packages/rabbitmq/src/amqp/connection.ts:421]: Should we catch errors in handlers and continue other handler execution if one fails?

File: packages/rabbitmq/src/rabbitmq.interfaces.ts
Comment (perf2711) [reply packages/rabbitmq/src/rabbitmq.interfaces.ts:49]: All right, but we should do something about empty messages. At this point, sending empty (i.e. empty byte array) to the queue will result in an JSON parsing error.

**Comment by WonderPanda** [comment]
Hey @perf2711 I pulled your work into a new branch so I could get it rebased on top of master and to add a few cleanup items. Closing this in favor of #134 there are a few small changes/additions to the work that you did but you'll still receive credit for all the contributions you've included this far

**Pull request closed without merging**
